File: gcp/frontend/src/services/capabilities.js
/**
 * Capabilities Service - Detects backend features based on available credentials.
 *
 * This service fetches capability information from the backend and caches it
 * for use throughout the frontend application.
 */

// Cached capabilities
let _capabilities = null;

/**
 * Fetch capabilities from the backend.
 *
 * @returns {Promise<Object>} Capabilities object with mode, features, models, limits
 */
export async function getCapabilities() {
    if (_capabilities !== null) {
        return _capabilities;
    }

    try {
        const response = await fetch('/api/capabilities/');
        if (!response.ok) {
            console.error('Failed to fetch capabilities:', response.status);
            return getDefaultCapabilities();
        }

        _capabilities = await response.json();
        console.log('Backend capabilities:', _capabilities);
        return _capabilities;
    } catch (error) {
        console.error('Error fetching capabilities:', error);
        return getDefaultCapabilities();
    }
}

/**
 * Get default (minimal) capabilities for when backend is unavailable.
 *
 * @returns {Object} Default capabilities object
 */
function getDefaultCapabilities() {
    return {
        mode: 'none',
        features: {
            chat: false,
            video_analysis: false,
            paper_rag: false,
            google_search: false
        },
        models: {
            chat: '',
            vision: ''
        },
        limits: {
            video_max_size_mb: 0,
            video_max_duration_sec: 0
        }
    };
}

/**
 * Check if a specific feature is enabled.
 *
 * @param {string} feature - Feature name (chat, video_analysis, paper_rag, google_search)
 * @returns {Promise<boolean>} Whether the feature is available
 */
export async function isFeatureEnabled(feature) {
    const caps = await getCapabilities();
    return caps?.features?.[feature] ?? false;
}

/**
 * Get the current mode (free, vertex, or none).
 *
 * @returns {Promise<string>} Current mode
 */
export async function getMode() {
    const caps = await getCapabilities();
    return caps?.mode ?? 'none';
}

/**
 * Check if running in Vertex AI mode.
 *
 * @returns {Promise<boolean>} Whether Vertex AI mode is active
 */
export async function isVertexMode() {
    const mode = await getMode();
    return mode === 'vertex';
}

/**
 * Get model names being used.
 *
 * @returns {Promise<Object>} Object with chat and vision model names
 */
export async function getModels() {
    const caps = await getCapabilities();
    return caps?.models ?? { chat: '', vision: '' };
}

/**
 * Reset cached capabilities (useful for testing or after mode change).
 */
export function resetCapabilitiesCache() {
    _capabilities = null;
}

/**
 * Get a mode badge string for display.
 *
 * @returns {Promise<string>} Badge text like "FREE" or "VERTEX AI"
 */
export async function getModeBadge() {
    const mode = await getMode();
    switch (mode) {
        case 'vertex':
            return 'VERTEX AI';
        case 'free':
            return 'FREE';
        default:
            return 'OFFLINE';
    }
}

/**
 * Get available tools based on capabilities.
 * Returns list of tool names that should be enabled in the chat.
 *
 * @returns {Promise<string[]>} Array of enabled tool names
 */
export async function getEnabledTools() {
    const caps = await getCapabilities();
    const tools = [];

    // Always available when chat is enabled
    if (caps.features.chat) {
        tools.push('load_trajectory');
        tools.push('navigate_experiment');
        tools.push('analyze_alignment');
    }

    // Vertex-only features
    if (caps.features.video_analysis) {
        tools.push('analyze_video');
    }

    if (caps.features.paper_rag) {
        tools.push('search_papers');
    }

    if (caps.features.google_search) {
        tools.push('web_search');
    }

    return tools;
}

File: gcp/web/api/capabilities.py
"""
Capabilities API - Exposes backend capabilities based on available credentials.

This endpoint allows the frontend to detect which features are available
and adjust the UI accordingly.
"""

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel
from services.vertex_client import ClientMode, get_vertex_client, reset_client

router = APIRouter()


class FeaturesResponse(BaseModel):
    """Available features."""

    chat: bool
    video_analysis: bool
    paper_rag: bool
    google_search: bool


class ModelsResponse(BaseModel):
    """Available models."""

    chat: str
    vision: str


class LimitsResponse(BaseModel):
    """Feature limits."""

    video_max_size_mb: int
    video_max_duration_sec: int


class CapabilitiesResponse(BaseModel):
    """Full capabilities response."""

    mode: str  # "free" or "vertex"
    features: FeaturesResponse
    models: ModelsResponse
    limits: LimitsResponse


@router.get("/", response_model=CapabilitiesResponse)
async def get_capabilities() -> CapabilitiesResponse:
    """
    Get backend capabilities based on available credentials.

    Returns information about:
    - Current mode (free or vertex)
    - Available features (chat, video, RAG, search)
    - Model names being used
    - Feature limits

    The frontend uses this to:
    - Show/hide features in the UI
    - Display mode indicator
    - Enforce limits (e.g., video size)
    """
    try:
        client = get_vertex_client()
        caps = client.capabilities

        return CapabilitiesResponse(
            mode=caps.mode.value,
            features=FeaturesResponse(
                chat=caps.chat,
                video_analysis=caps.video_analysis,
                paper_rag=caps.paper_rag,
                google_search=caps.google_search,
            ),
            models=ModelsResponse(
                chat=caps.chat_model,
                vision=caps.vision_model,
            ),
            limits=LimitsResponse(
                video_max_size_mb=caps.video_max_size_mb,
                video_max_duration_sec=caps.video_max_duration_sec,
            ),
        )
    except ValueError:
        # No credentials configured - return minimal response
        return CapabilitiesResponse(
            mode="none",
            features=FeaturesResponse(
                chat=False,
                video_analysis=False,
                paper_rag=False,
                google_search=False,
            ),
            models=ModelsResponse(
                chat="",
                vision="",
            ),
            limits=LimitsResponse(
                video_max_size_mb=0,
                video_max_duration_sec=0,
            ),
        )


@router.get("/health")
async def health_check() -> dict[str, Any]:
    """
    Health check endpoint.

    Returns basic status and mode information.
    """
    try:
        client = get_vertex_client()
        return {
            "status": "healthy",
            "mode": client.mode.value,
            "project": client.capabilities.mode == ClientMode.VERTEX,
        }
    except ValueError:
        return {
            "status": "unconfigured",
            "mode": "none",
            "project": False,
        }


@router.post("/reset")
async def reset_client_endpoint() -> dict[str, str]:
    """
    Reset the client (useful for development/testing).

    This forces the client to re-detect credentials on next request.
    """
    reset_client()
    return {"status": "reset"}
